File: Photo Picker ObjC/Photo Picker ObjC/ViewController.m
//
//  ViewController.m
//  Photo Picker ObjC
//
//  Created by Kyle Howells on 23/06/2020.
//

#import "ViewController.h"
#import <Photos/Photos.h>
#import <PhotosUI/PhotosUI.h>
@import MobileCoreServices;

@interface ViewController () <PHPickerViewControllerDelegate>

@end

@implementation ViewController{
	UIScrollView *scrollView;
	NSMutableArray <UIImageView*>* imageViews;
	UIButton *selectButton;
}

- (void)viewDidLoad {
	[super viewDidLoad];
	
	self.view.backgroundColor = [UIColor whiteColor];
	
	imageViews = [NSMutableArray array];
	
	// Create ScrollView
	scrollView = [[UIScrollView alloc] init];
	scrollView.backgroundColor = [UIColor colorWithWhite:0.98 alpha:1];
	[self.view addSubview:scrollView];
	
	// Select Photos Button
	selectButton = [UIButton buttonWithType:UIButtonTypeSystem];
	[selectButton setTitle:@"Select Photos" forState:UIControlStateNormal];
	[selectButton addTarget:self action:@selector(selectPressed:) forControlEvents:UIControlEventTouchUpInside];
	[selectButton sizeToFit];
	[self.view addSubview:selectButton];
}

-(void)viewDidLayoutSubviews{
	[super viewDidLayoutSubviews];
	const CGSize size = self.view.bounds.size;
	const UIEdgeInsets safeArea = self.view.safeAreaInsets;
	
	selectButton.frame = ({
		CGRect frame = CGRectZero;
		frame.size.width = MIN(size.width - 20, 250);
		frame.size.height = 40;
		frame.origin.y = size.height - (frame.size.height + 20 + safeArea.bottom);
		frame.origin.x = (size.width - frame.size.width) * 0.5;
		frame;
	});
	
	scrollView.frame = ({
		CGRect frame = CGRectZero;
		frame.origin.y = safeArea.top + 10;
		frame.size.height = (selectButton.frame.origin.y - 20) - frame.origin.y;
		frame.size.width = size.width - 40;
		frame.origin.x = (size.width - frame.size.width) * 0.5;
		frame;
	});
	
	CGFloat y = 10;
	for (NSInteger i = 0; i < imageViews.count; i++) {
		UIImageView *imageView = imageViews[i];
		imageView.frame = ({
			CGRect frame = CGRectZero;
			frame.origin.y = y;
			frame.size.width = MIN(scrollView.bounds.size.width - 20, 300);
			frame.origin.x = (scrollView.bounds.size.width - frame.size.width) * 0.5;
			frame.size.height = MIN(frame.size.width * 0.75, 250);
			y += frame.size.height + 10;
			frame;
		});
	}
	scrollView.contentSize = CGSizeMake(0, y);
}

#pragma mark Helpers

-(UIImageView*)newImageViewForImage:(UIImage*)image{
	UIImageView *imageView = [[UIImageView alloc] init];
	imageView.contentMode = UIViewContentModeScaleAspectFit;
	imageView.backgroundColor = [UIColor blackColor];
	imageView.image = image;
	return imageView;
}

-(void)clearImageViews{
	for (UIImageView *imageView in imageViews) {
		[imageView removeFromSuperview];
	}
	[imageViews removeAllObjects];
}

#pragma mark - PHPicker

-(void)selectPressed:(id)sender{
	PHPickerConfiguration *config = [[PHPickerConfiguration alloc] init];
	config.selectionLimit = 3;
	config.filter = [PHPickerFilter imagesFilter];
	
	PHPickerViewController *pickerViewController = [[PHPickerViewController alloc] initWithConfiguration:config];
	pickerViewController.delegate = self;
	[self presentViewController:pickerViewController animated:YES completion:nil];
}

-(void)picker:(PHPickerViewController *)picker didFinishPicking:(NSArray<PHPickerResult *> *)results{
	NSLog(@"-picker:%@ didFinishPicking:%@", picker, results);
	
	[self clearImageViews];
	[picker dismissViewControllerAnimated:YES completion:nil];
	
	for (PHPickerResult *result in results) {
		NSLog(@"result: %@", result);
		
		NSLog(@"%@", result.assetIdentifier);
		NSLog(@"%@", result.itemProvider);
		
		// Get UIImage
		[result.itemProvider loadObjectOfClass:[UIImage class] completionHandler:^(__kindof id<NSItemProviderReading>  _Nullable object, NSError * _Nullable error) {
			NSLog(@"object: %@, error: %@", object, error);

			if ([object isKindOfClass:[UIImage class]]) {
				dispatch_async(dispatch_get_main_queue(), ^{
					UIImageView *imageView = [self newImageViewForImage:(UIImage*)object];
					[self->imageViews addObject:imageView];
					[self->scrollView addSubview:imageView];
					[self.view setNeedsLayout];
				});
			}
		}];
		
		// Get file
//		[result.itemProvider loadItemForTypeIdentifier:(NSString *)kUTTypeImage options:nil completionHandler:^(id item, NSError *error) {
//			NSLog(@"%@", item);
//			NSLog(@"%@", [item class]);
//
//			if ([item isKindOfClass:[NSURL class]]) {
//				NSError *error = nil;
//				NSData *data = [NSData dataWithContentsOfURL:item options:0 error:&error];
//				NSLog(@"%@", data);
//				NSLog(@"%@", error);
//			}
//		}];
	}
}

@end

File: Photo Picker Swift/Photo Picker Swift/ViewController.swift
//
//  ViewController.swift
//  Photo Picker Swift
//
//  Created by Kyle Howells on 23/06/2020.
//

import UIKit
import PhotosUI

class ViewController: UIViewController, PHPickerViewControllerDelegate {

	// MARK: - PHPickerViewController
	
	@objc func pickPhotos()
	{
		var config = PHPickerConfiguration()
		config.selectionLimit = 3
		config.filter = PHPickerFilter.images
		
		let pickerViewController = PHPickerViewController(configuration: config)
		pickerViewController.delegate = self
		self.present(pickerViewController, animated: true, completion: nil)
	}
	
	// MARK: PHPickerViewControllerDelegate
	
	func picker(_ picker: PHPickerViewController, didFinishPicking results: [PHPickerResult]) {
		picker.dismiss(animated: true, completion: nil)
		print(picker)
		print(results)
		
		for result in results {
			print(result.assetIdentifier)
			print(result.itemProvider)
			
			result.itemProvider.loadObject(ofClass: UIImage.self, completionHandler: { (object, error) in
				print(object)
				print(error)
				if let image = object as? UIImage {
					DispatchQueue.main.async {
						let imv = self.newImageView(image: image)
						self.imageViews.append(imv)
						self.scrollView.addSubview(imv)
						self.view.setNeedsLayout()
					}
				}
			})
		}
	}
	
	
	
	// MARK: - View Setup
	
	lazy var scrollView:UIScrollView = {
		let s = UIScrollView()
		s.backgroundColor = UIColor(white: 0.98, alpha: 1)
		return s
	}()
	
	lazy var button:UIButton = {
		let b = UIButton(type: .system)
		b.setTitle("Select Photos", for: .normal)
		b.addTarget(self, action: #selector(pickPhotos), for: .touchUpInside)
		b.sizeToFit()
		return b
	}()
	
	var imageViews = [UIImageView]()
	
	func newImageView(image:UIImage?) -> UIImageView {
		let imv = UIImageView()
		imv.backgroundColor = .black
		imv.image = image
		return imv
	}
	
	override func viewDidLoad() {
		super.viewDidLoad()
		
		self.view.backgroundColor = UIColor.white
		
		self.view.addSubview(self.scrollView)
		self.view.addSubview(self.button)
	}
	
	override func viewDidLayoutSubviews() {
		super.viewDidLayoutSubviews()
		let size = self.view.bounds.size
		let safeArea = self.view.safeAreaInsets
		let padding:CGFloat = 10
		
		button.frame = {
			var f = CGRect.zero
			f.size.width = min(size.width - (padding * 2), 250)
			f.size.height = 40
			f.origin.x = (size.width - f.width) * 0.5
			f.origin.y = size.height - (safeArea.bottom + padding + f.size.height)
			return f
		}()
		
		scrollView.frame = {
			var f = CGRect.zero
			f.origin.y = safeArea.top + padding
			f.size.width = size.width - (padding * 2)
			f.size.height = (button.frame.minY - 20) - f.origin.y
			f.origin.x = (size.width - f.width) * 0.5
			return f
		}()
		
		var y:CGFloat = 10
		for imageView in imageViews {
			imageView.frame = {
				var f = CGRect.zero
				f.origin.y = y
				f.size.width = min(scrollView.bounds.width - (padding * 2), 300)
				f.size.height = min(f.width * 0.75, 250)
				f.origin.x = (scrollView.bounds.width - f.size.width) * 0.5
				y += f.size.height + padding
				return f
			}()
		}
		scrollView.contentSize = CGSize(width: 0, height: y)
	}
}
